### PR TITLE
Fix route sending of subs list and removing fan-in fix

### PR DESCRIPTION
**Pull request by kozlovic**
Would like to merge those 2 fixes in the 1.4.x branch.

[FIXED] Route may disconnect when sending subscription interest
[FIXED] Removing fan-in fix

/cc @nats-io/core

### CODE REVIEW

File: server/route.go [hunk 1]
```diff
@@ -584,24 +587,32 @@ func (s *Server) sendLocalSubsToRoute(route *client) {
...
 		// Send SUB interest only if subject has a match in import permissions
 		if !route.canImport(sub.subject) {
 			continue
 		}
 		proto := fmt.Sprintf(subProto, sub.subject, sub.queue, routeSid(sub))
 		route.queueOutbound([]byte(proto))
-		if route.out.pb > int64(route.out.sz*2) {
-			route.flushSignal()
+		if route.out.pb > threshold {
+			route.flushOutbound()
```
Comment (derekcollison) [inline server/route.go]: Should check to see if this actually did anything and if not do a signal.

File: server/client.go [hunk 2]
```diff
@@ -477,7 +477,7 @@ func (c *client) readLoop() {
 		// Budget to spend in place flushing outbound data.
 		// Client will be checked on several fronts to see
 		// if applicable. Routes will never wait in place.
-		budget := time.Millisecond
+		budget := 500 * time.Microsecond
```
Comment (derekcollison) [inline server/client.go]: Why the change here?

Comment (kozlovic) [reply server/client.go]: Because this change was part of the fan-in fix. If I revert the fix, I want to restore the code as close as possible to what it was prior to the fix.

Comment (derekcollison) [reply server/client.go]: ok gotcha.

File: server/route.go (hunk 1 above)
Comment (kozlovic) [reply server/route.go]: not needed because it returns false only in one condition: the flushOutbound flag is already set and in that case the lock is not released, and then we signal at the end of this loop.

Comment (derekcollison) [reply server/route.go]: It could be released, but if you signal at the end of the loop and you don't think you need a kick in between then its ok by me.

**Comment by derekcollison** [comment]
LGTM

File: server/route.go (hunk 1 above)
Comment (kozlovic) [reply server/route.go]: Actually I think we should do 2 things: signal and release/reacquire lock. If we don't, here is what is happening. Say we are in this loop and call flush because we have accumulated some data. Another go routine is in flush (and the flag is set) and has released the lock to do the actual socket write. If we don't signal/release lock/reacquire lock, then this loop will continue to enqueue and every time it calls flushOutbound() it would get false: the other go-routine will not be able to grab the lock after the socket write.
The point is that simply signaling is not going to help if we don't release the lock.

Comment (derekcollison) [reply server/route.go]: Agree.

Comment (kozlovic) [reply server/route.go]: will calling runtime.Gosched() in between be ok or do you prefer a very small sleep (if so, what value do you think is appropriate)?

Comment (derekcollison) [reply server/route.go]: If you release the lock that will engage the scheduler. No need IMO for explicit call or a sleep.

File: server/client.go [hunk 3]
```diff
@@ -181,9 +181,9 @@ type outbound struct {
 	pb  int64         // Total pending/queued bytes.
 	pm  int64         // Total pending/queued messages.
 	sg  *sync.Cond    // Flusher conditional for signaling.
-	wdl time.Duration // Snapshot fo write deadline.
-	mp  int64         // snapshot of max pending.
 	fsp int           // Flush signals that are pending from readLoop's pcd.
+	mp  int64         // snapshot of max pending.
+	wdl time.Duration // Snapshot fo write deadline.
```
Comment (wallyqs) [inline server/client.go]:
```suggestion
	wdl time.Duration // Snapshot of write deadline.
```

File: server/client.go [hunk 4]
```diff
@@ -181,9 +181,9 @@ type outbound struct {
 	pb  int64         // Total pending/queued bytes.
 	pm  int64         // Total pending/queued messages.
 	sg  *sync.Cond    // Flusher conditional for signaling.
-	wdl time.Duration // Snapshot fo write deadline.
-	mp  int64         // snapshot of max pending.
 	fsp int           // Flush signals that are pending from readLoop's pcd.
+	mp  int64         // snapshot of max pending.
```
Comment (wallyqs) [inline server/client.go]:
```suggestion
	mp  int64         // Snapshot of max pending.
```

**Comment by kozlovic** [comment]
@derekcollison @wallyqs Thank you both for the comments. I have updated each commit with the appropriate fix and push force.
